pkg/store/bucket.go

```diff
@@ -996,44 +996,43 @@ func int64index(s []int64, x int64) int {
 }
 
 // getFor returns a time-ordered list of blocks that cover date between mint and maxt.
-// Blocks with the lowest resolution possible but not lower than the given resolution are returned.
-func (s *bucketBlockSet) getFor(mint, maxt, minResolution int64) (bs []*bucketBlock) {
-	if mint == maxt {
-		return nil
-	}
-
+// The covering is greedy, preferring blocks of the lowest allowed resolutions.
+func (s *bucketBlockSet) getFor(minTime, maxTime, minResolution, maxResolution int64) (bs []*bucketBlock) {
 	s.mtx.RLock()
 	defer s.mtx.RUnlock()
 
-	// Find first matching resolution.
-	i := 0
-	for ; i < len(s.resolutions) && s.resolutions[i] > minResolution; i++ {
+	for i := 0; i < len(s.resolutions); i++ {
+		if s.resolutions[i] >= minResolution && s.resolutions[i] <= maxResolution {
+			if len(bs) == 0 {
+				// Nothing covered so far, cover as much as possible of whole range with current resolution.
+				bs = coverInterval(minTime, maxTime, s.blocks[i])
+			} else {
```

Comment:
As general tip: I would change `else` to just continue on above case, bit cleaner.